examples/communication/grandchild_to_grandparent/src/main.rs

```diff
@@ -0,0 +1,145 @@
+use std::rc::Rc;
+use yew::{html, Callback, Component, Context, ContextHandle, ContextProvider, Html};
+
+// This is the shared state between the parent and child components.
+#[derive(Clone, PartialEq)]
+pub struct AppState {
+    value: u32,
+    child_clicked: Callback<()>,
+}
+
+// Our top-level component.
+
+pub struct GrandParent {
+    state: Rc<AppState>,
+}
+
+pub enum Msg {
+    ButtonClick,
+}
+
+impl Component for GrandParent {
+    type Message = Msg;
+    type Properties = ();
+
+    fn create(ctx: &Context<Self>) -> Self {
+        let child_clicked = ctx.link().callback(|_| Msg::ButtonClick);
+        let state = Rc::new(AppState {
+            value: 0,
+            child_clicked,
+        });
+        Self { state }
+    }
+
+    fn update(&mut self, _ctx: &Context<Self>, msg: Self::Message) -> bool {
+        match msg {
+            Msg::ButtonClick => {
+                let value = self.state.value + 1;
+                self.state = Rc::new(AppState {
+                    value,
+                    ..(*self.state).clone()
+                });
+                true
+            }
+        }
+    }
+
+    fn view(&self, _ctx: &Context<Self>) -> Html {
+        let app_state = self.state.clone();
+        let msg = format!(
+            "My grandchildren have been clicked {} times",
+            self.state.value
+        );
+
+        html! {
+            <ContextProvider<Rc<AppState>> context={app_state}>
+                <div class="app">
+                    <div class="parent">
+                        <h2>{ "Grandchild-to-Grandparent (and back!) communication example" }</h2>
+                        <div>{msg}</div>
+                        <Parent />
+                    </div>
+                </div>
+            </ContextProvider<Rc<AppState>>>
+        }
+    }
+}
+
+// The `Parent` component is the parent of the `Child` component. It has no logic, and is here to show there is no direct
+// relation between grandchild and grandparent.
```

Comment:
Should be a doc comment. Also, I think this component is simple enough to be a function component again.